doozerlib/cli/release_gen_payload.py

```diff
@@ -58,762 +145,869 @@ def modify_and_replace_api_object(api_obj: oc.APIObject, modifier_func: Callable
-                    # we've calculated a sha256 of all the manifests being added.
-                    'image': f'quay.io/{organization}/{repository}:sha256-{overall_manifest_hash.hexdigest()}',
-                    'manifests': manifests
-                }
-
-                with component_manifest_path.open(mode='w+') as ml:
-                    yaml.safe_dump(ml_dict, stream=ml, default_flow_style=False)
-
-                output_pullspec = ml_dict['image']
-                exectools.cmd_assert(f'manifest-tool push from-spec {str(component_manifest_path)}', retries=3)
-                # if we are actually pushing a manifest list, then we should derive a sha256 based pullspec
-                output_registry_org_repo = output_pullspec.rsplit(':')[0]  # e.g. quay.io/openshift-release-dev/ocp-v4.0-art-dev:sha256-b056..84b-ml -> quay.io/openshift-release-dev/ocp-v4.0-art-dev
-                output_digest_pullspec = output_registry_org_repo + '@' + find_manifest_list_sha(output_pullspec)  # create a sha based pullspec for the new manifest list
-
-            multi_istags.append(PayloadGenerator.build_payload_istag(tag_name, PayloadGenerator.PayloadEntry(
+        return multi_release_name, multi_release_manifest_list_tag
+
+    def build_multi_istag(self, tag_name: str, arch_to_payload_entry: Dict[str, PayloadEntry], imagestream_namespace: str) -> Dict:
+        """Build a single imagestream tag for a component in a multi-arch payload."""
+        # There are two flows:
+        # 1. The images for ALL arches were part of the same brew built manifest list. In this case,
+        #    we want to reuse the manifest list (it was already mirrored during the mirroring step).
+        # 2. At least one arch for this component does not have the same manifest list as the
+        #    other images. This will always be true for RHCOS, but also applies
+        #    to -alt images. In this case, we must stitch a manifest list together ourselves.
+        entries: List[PayloadEntry] = list(arch_to_payload_entry.values())
+        # Determine which flow applies by checking whether all entries have the same manifest list
+        manifest_list_dests: Set[str] = set(entry.dest_manifest_list_pullspec for entry in entries)
+        if len(manifest_list_dests) == 1 and set(manifest_list_dests).pop():  # only one, and not None
+            # Flow 1: Just reuse the manifest list built in brew and synced to a tag in quay.
+            output_digest_pullspec = exchange_pullspec_tag_for_shasum(
+                manifest_list_dests.pop(),
+                entries[0].build_inspector.get_manifest_list_digest())
+            self.logger.info(f"Reusing brew manifest-list {output_digest_pullspec} for component {tag_name}")
+        else:
+            # Flow 2: Build a new manifest list and push it to quay.
+            output_digest_pullspec = self.create_multi_manifest_list(tag_name, arch_to_payload_entry, imagestream_namespace)
```

Comment:
I think we'll also get this flow for those that don't build for all arches, since `pod` will be subbed in for non-building arches. But what if we just used the original manifest list for those too? The reference would be satisfied but never used in the non-building arches, shouldn't that be fine?

Reply:
I feel there is a chance we'd end up breaking something in the ecosystem. Historical releases have guaranteed that if container tooling queries or pulls a release tag, there will be _something_ there for its arch. In the proposed approach, an s390x client may have a manifest list reference to try to inspect/pull, but an error would result if it tried. 

The only reward I can think of for taking the risk is avoiding the creation of several manifest lists in ART's pipeline. If you believe that is a significant enough time savings, perhaps we could turn on the strategy for 4.12 only and see if anything breaks.

Reply:
(after discussion)
The point is well taken, nothing should _need_ to look at the missing arch images, but something might be anyway and could break when its expectations aren't met.

My concern was more that we're going to build new manifest lists for these every time we sync, which I'd like to avoid just for hygiene. You came up with a good solution - checksum the manifest list content and tag the manifest list in quay per that checksum; then when a manifest list is called for, look for it under the checksum before creating a new one. (And just for the sake of security, check that the manifest list content does indeed match the desired content, and explode if it doesn't because quay is compromised in that case.)

I think this will be a follow-on improvement.